timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java

```diff
@@ -124,83 +130,103 @@ private void deleteOldIndex(String currIndex) throws IOException {
+            if (responses.hasFailures()) {
+                LOG.error("Couldn't bulk index tasks to elasticsearch cluster. Error: {}", responses.buildFailureMessage());
+            }
         } catch (IOException e) {
-            LOG.error("Could not delete index " + index, e);
+            LOG.error("Couldn't bulk index tasks to elasticsearch cluster.", e);
         }
-        //TODO not correct - Should be changed to Rollover
 
+    }
+
+    void index(Map<String, Task> tasksMap) {
+        BulkRequest request = new BulkRequest();
+        int currBatch = 0;
+
+        Collection<UpdateRequest> requests = new ArrayList<>();
+        for (Map.Entry<String, Task> taskEntry : tasksMap.entrySet()) {
+            Task task = taskEntry.getValue();
+
+            String index = getTaskIndexWithEnv(TIMBERMILL_INDEX_PREFIX, task.getEnv(), ZonedDateTime.now());
+            try {
+                deleteOldIndex(index);
+            } catch (IOException e) {
+                LOG.error("Could not delete index " + index, e);
+            }
+            //TODO not correct - Should be changed to Rollover
 
-        List<UpdateRequest> requests = new ArrayList<>();
-        Map<String, List<Event>> eventsMap = events.stream().collect(groupingBy(e -> e.getTaskId()));
-        for (Map.Entry<String, List<Event>> eventsEntry : eventsMap.entrySet()) {
-            Task task = new Task(eventsEntry.getValue());
-            UpdateRequest updateRequest = task.getUpdateRequest(index, eventsEntry.getKey());
+            UpdateRequest updateRequest = task.getUpdateRequest(index, taskEntry.getKey());
             requests.add(updateRequest);
         }
 
+        int currentSize = 0;
+        Collection<Future> futures = new ArrayList<>();
         for (UpdateRequest updateRequest : requests) {
-            i++;
-            currBatch++;
-
             request.add(updateRequest);
-            try{
-                if (((i % bulkSize) == 0) || (i == requests.size())) {
-                    BulkResponse responses = client.bulk(request, RequestOptions.DEFAULT);
-
-                    if (responses.hasFailures()) {
-                        retryUpdateWithSmallerBulkSizeOrFail(events, bulkSize, tryNum, responses.buildFailureMessage());
-                    }
-                    LOG.info("Batch of {} tasks indexed successfully", currBatch);
-                    currBatch = 0;
-                    request = new BulkRequest();
-                }
-            } catch (IOException e) {
-                retryUpdateWithSmallerBulkSizeOrFail(events, bulkSize, tryNum, e.getMessage());
+            currentSize += request.estimatedSizeInBytes();
+
+            if (currentSize > indexBulkSize) {
+                BulkRequest finalRequest = request;
+                futures.add(executorService.submit(() -> sendBulkRequest(finalRequest)));
+                LOG.info("Batch of {} tasks indexed successfully", currBatch);
```

Comment:
isn't that confusing since it's a future? maybe change to "batch # sent succsefully"?